src/main/java/org/ice4j/ice/Agent.java

```diff
@@ -2298,6 +2298,17 @@ public void free()
         if (stunKeepAliveThread != null)
             stunKeepAliveThread.interrupt();
 
+        // cancel termination timer in case agent is freed
+        // before termination timer is triggered
+        synchronized (terminationThreadSyncRoot)
```

Comment:
>> If the thread is still alive, it is most probably in the call to wait. If that's the case, interrupting it will result in it executing terminate(TERMINATED), which is probably not what we want.

Yes, but I didn't consider it as serious issue, since Agent is anyway freeing. Actually I have a better version of this fix which does not have this problem, but I wanted to start with smaller pull request.

>> What is the problem you're trying to solve? This doesn't solve the problem of terminate executing after the agent is freed, and the leak should only last 3 seconds at most (the default termination delay).

The problem I'm trying to solve is described in my [question](https://community.jitsi.org/t/webrtc-client-to-jvb-connection-recovery/15418) for Scenario 2.
I basically set delay timeout to "infinite" value via configuration to effectively cancel termination, it caused thread to live "forever" in waiting state and it probably also keep reference to dead Agent.

I'll update this pull request with better version of fix, which does proper cancellation as well as saves a lot of threads for my scenario by using scheduled thread pool execution service to execute delayed termination.

Reply:
I'd say that for the infinite termination feature this thread should not be started at all.

Reply:
By "infinite" I meant something like day or week (I believe no peer will connect longer in single session) :) But we could also define some value as infinite and do not start delayed termination task at all. BTW, there is more recent version of this pull request which does not create threads explicitly at all.